flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -192,6 +211,13 @@ public void registerAll(
     @Override
     public void registerAllAfterRestored(CompletedCheckpoint checkpoint, RestoreMode mode) {
         registerAll(checkpoint.getOperatorStates().values(), checkpoint.getCheckpointID());
+        checkpoint
+                .getRestoredProperties()
```

Comment:
When off->on, if the "off" checkpoint is V3, the following branch will not be executed:
```Java
 if (entry.preventDiscardingCreatedCheckpoint
                && restoredCheckpointSharingStrategies.containsKey(entry.createdByCheckpointID)) {
            return true;
        }
```

Reply:
Good catch! I've changed the type of map values to `Optional`.